util/include/Metrics.hpp

```diff
@@ -0,0 +1,186 @@
+// Concord
+//
+// Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#ifndef CONCORD_BFT_METRICS_HPP
+#define CONCORD_BFT_METRICS_HPP
+
+#include <stdint.h>
+#include <map>
+#include <vector>
+#include <mutex>
+#include <memory>
+
+namespace concordMetrics {
+
+// Forward declarations since Aggegator requires these types.
+class Component;
+class Values;
+class Gauge;
+class Status;
+class Counter;
+
+// An aggregator maintains metrics for multiple components. Components
+// maintain a handle to the aggregator and update it periodically with
+// all their metric values. Therefore, the state of all metrics is eventually
+// consistent.
+//
+// The Aggregator is the type responsible for reporting metrics for the entire
+// system. A process should have a single aggregator, and any service
+// responsible for reporting system metrics should read it from the aggregator.
+class Aggregator {
+ public:
+  Gauge GetGauge(const std::string& component_name,
+                 const std::string& val_name);
+  Status GetStatus(const std::string& component_name,
+                   const std::string& val_name);
+  Counter GetCounter(const std::string& component_name,
+                     const std::string& val_name);
+
+ private:
+  void RegisterComponent(Component& component);
+  void UpdateValues(const std::string& name, Values&& values);
+
+  std::map<std::string, Component> components_;
+  std::mutex lock_;
+
+  friend class Component;
+};
+
+// A Gauge is a an integer value that shows the current value of something. It
+// can only be varied by directly setting and getting it.
+class Gauge {
+ public:
+  explicit Gauge(const uint64_t val) : val_(val) {}
+
+  void Set(const uint64_t val) { val_ = val; }
+  uint64_t Get() { return val_; }
+
+ private:
+  uint64_t val_;
+};
+
+// Status is a text based representation of a value. It's used for things that
+// don't have strictly numeric representations, like the current state of the
+// BFT or the last message received.
+class Status {
```

Comment:
I'd prefer not to here, mainly because the api is only incidentally the same. They are logically separate types and potentially can evolve to be more different.